src/pudl/etl/epacems_assets.py

```diff
@@ -53,30 +67,80 @@ def hourly_emissions_epacems(
     )
     partitioned_path.mkdir(exist_ok=True)
+
+    for state in epacems_settings.states:
+        logger.info(f"Processing EPA CEMS hourly data for {year}-{state}")
+        df = pudl.extract.epacems.extract(year=year, state=state, ds=ds)
+        df = pudl.transform.epacems.transform(df, epacamd_eia, plants_entity_eia)
+        table = pa.Table.from_pandas(df, schema=schema, preserve_index=False)
+
+        # Write to a directory of partitioned parquet files
+        with pq.ParquetWriter(
+            where=partitioned_path / f"epacems-{year}-{state}.parquet",
+            schema=schema,
+            compression="snappy",
+            version="2.6",
+        ) as partitioned_writer:
+            partitioned_writer.write_table(table)
+
+    return YearPartitions(year, epacems_settings.states)
+
+
+@op(
+    config_schema={
+        "pudl_output_path": Field(
+            EnvVar(
+                env_var="PUDL_OUTPUT",
+            ),
+            description="Path of directory to store the database in.",
+            default_value=None,
+        ),
+    },
+)
+def consolidate_partitions(context, partitions: list[YearPartitions]) -> None:
+    """Read partitions into memory and write to a single monolithic output.
+
+    Args:
+        context: dagster keyword that provides access to resources and config.
+        partitions: Year and state combinations in the output database.
+    """
+    partitioned_path = (
+        Path(context.op_config["pudl_output_path"]) / "hourly_emissions_epacems"
+    )
     monolithic_path = (
         Path(context.op_config["pudl_output_path"]) / "hourly_emissions_epacems.parquet"
     )
+    schema = Resource.from_id("hourly_emissions_epacems").to_pyarrow()
 
     with pq.ParquetWriter(
         where=monolithic_path, schema=schema, compression="snappy", version="2.6"
     ) as monolithic_writer:
-        for part in epacems_settings.partitions:
-            year = part["year"]
-            state = part["state"]
-            logger.info(f"Processing EPA CEMS hourly data for {year}-{state}")
-            df = pudl.extract.epacems.extract(year=year, state=state, ds=ds)
-            df = pudl.transform.epacems.transform(df, epacamd_eia, plants_entity_eia)
-            table = pa.Table.from_pandas(df, schema=schema, preserve_index=False)
-
-            # Write to one monolithic parquet file
-            monolithic_writer.write_table(table)
-
-            if context.op_config["partition"]:
-                # Write to a directory of partitioned parquet files
-                with pq.ParquetWriter(
-                    where=partitioned_path / f"epacems-{year}-{state}.parquet",
-                    schema=schema,
-                    compression="snappy",
-                    version="2.6",
-                ) as partitioned_writer:
-                    partitioned_writer.write_table(table)
+        for year, states in partitions:
+            for state in states:
+                monolithic_writer.write_table(
```

Comment:
Thanks for the tip on the pyarrow implementation, @zaneselvans this is much quicker than the previous way I was trying to do this!